hal/targets/hal/TARGET_ONSEMI/TARGET_NCS36510/gpio_api.c

```diff
@@ -0,0 +1,195 @@
+
+/* Include from the mbed-hal layer */
+#include "gpio_api.h" 
+#include "pinmap.h"
+
+
+/** Set the given pin as GPIO
+ *
+ * @param pin The pin to be set as GPIO
+ * @return The GPIO port mask for this pin
+ **/
+uint32_t gpio_set(PinName pin)
+{
+  if (pin != NC)
+  {
+    	/* Configure to GPIO using pin function API*/
+        pin_function(pin, CONFIGURE_AS_GPIO);
+      
+        return ((uint32_t) 0x1 << pin);
+        
+  } 
+  
+  return(0x00000000);
+  
+}
+
+/** Initialize the GPIO pin
+ *
+ * @param obj The GPIO object to initialize
+ * @param pin The GPIO pin to initialize
+ */
+void gpio_init(gpio_t *obj, PinName pin)
+{
+  /* Initialize the GPIO membase */
+  obj->GPIOMEMBASE = GPIOREG;
+  
+  /* Initialize the pin to be GPIO */
+  obj->gpioPin = pin;
+  obj->gpioMask = gpio_set(pin);  
+  
+  /* Enable the GPIO clock */
+  CLOCK_ENABLE(CLOCK_GPIO);  
+  
+  /* Set the drive strength of the pin to 1 by default */
+  /** - Get PAD IO register address for the PAD number */
+  PadReg_t *PadRegOffset = (PadReg_t*)(PADREG_BASE + (pin * PAD_REG_ADRS_BYTE_SIZE));
+
+  /* - Disable the GPIO clock */
+  CLOCK_DISABLE(CLOCK_GPIO);
+
+  /** - Enable the clock for PAD peripheral device */
+  CLOCK_ENABLE(CLOCK_PAD);
+  
+  /** - Set drive type, pulltype & drive strength */
+  PadRegOffset->PADIO0.BITS.POWER = 1;
+  
+  /** - Disable the clock for PAD peripheral device */
+  CLOCK_DISABLE(CLOCK_PAD); 
+}
+
+/** Set the input pin mode
+ *
+ * @param obj  The GPIO object
+ * @param mode The pin mode to be set
+ */
+void gpio_mode(gpio_t *obj, PinMode mode)
+{
+  uint32_t pin = obj->gpioPin;
+  
+  /* Set the mode for the pin */
+  pin_mode((PinName)pin, mode);
+}
+
+/** Set the pin direction
+ *
+ * @param obj       The GPIO object
+ * @param direction The pin direction to be set
+ */
+void gpio_dir(gpio_t *obj, PinDirection direction)
+{
+  /* Enable the GPIO clock */
+  CLOCK_ENABLE(CLOCK_GPIO);
+
+  if (direction == PIN_INPUT) {
+    obj->GPIOMEMBASE->W_IN = obj->gpioMask;
+  } else if (direction == PIN_OUTPUT) {
+    obj->GPIOMEMBASE->W_OUT = obj->gpioMask;
+  }
+
+  /* - Disable the GPIO clock */
+  CLOCK_DISABLE(CLOCK_GPIO);
+}
+
+/** Set the output value
+ *
+ * @param obj   The GPIO object
+ * @param value The value to be set
+ */
+void gpio_write(gpio_t *obj, int value)
+{
+  /* Enable the GPIO clock */
+  CLOCK_ENABLE(CLOCK_GPIO);
+
+  /* Set the GPIO based on value */
+  if (value)
+  {
+    obj->GPIOMEMBASE->R_STATE_W_SET = obj->gpioMask;
+  }
+  else
+  {
+    obj->GPIOMEMBASE->R_IRQ_W_CLEAR = obj->gpioMask;
+  }
+
+  /* - Disable the GPIO clock */
+  CLOCK_DISABLE(CLOCK_GPIO);
+}
+
+/** Read the input value
+ *
+ * @param obj The GPIO object
+ * @return An integer value 1 or 0
+ */
+int gpio_read(gpio_t *obj)
+{
+  int ret;
+
+  /* Enable the GPIO clock */
+  CLOCK_ENABLE(CLOCK_GPIO);
+
+  ret = (obj->GPIOMEMBASE->R_STATE_W_SET & obj->gpioMask) ? 1: 0;
+
+  /* - Disable the GPIO clock */
+  CLOCK_DISABLE(CLOCK_GPIO);
```

Comment:
Shouldnt you check if gpio clock is enabled? Assuming a user or any other external lib uses gpio, and do some clocking around it, functions like this disable clock withotu even considering the clock might have been enabled and used by something else?